edx_rest_framework_extensions/auth/jwt/middleware.py

```diff
@@ -258,19 +261,56 @@ def process_view(self, request, view_func, view_args, view_kwargs):  # pylint: d
         # .. custom_attribute_name: has_jwt_cookie
         # .. custom_attribute_description: Enables us to see requests which have the full reconstituted
         #      JWT cookie. If this attribute is missing, it is assumed to be False.
-        monitoring.set_custom_attribute('has_jwt_cookie', has_reconstituted_jwt_cookie)
-
+        set_custom_attribute('has_jwt_cookie', has_reconstituted_jwt_cookie)
+
+        # DRF authentication does not set the request.user early enough for it to be used in process_request/
+        # process_view of middleware. This code enables JWT cookie authentication to set the request.user for
+        # middleware, before it will presumably happen again during DRF authentication.
+        # For more info on DRF and the request.user, see
+        # https://github.com/jpadilla/django-rest-framework-jwt/issues/45#issuecomment-74996698
+        #
+        # If the user has already been authenticated, we already have a user and intentionally avoid resetting it. As of
+        # Oct 2023, this would likely be due to session authentication in middleware. It is thus important for
+        # JwtAuthentication to verify that the session user and JWT user match. It is possible that this would be better
+        # handled through a more traditional AuthenticationMiddleware that handles both JWT cookies and sessions in
+        # the future.
         if has_reconstituted_jwt_cookie and get_setting(ENABLE_SET_REQUEST_USER_FOR_JWT_COOKIE):
-            # DRF does not set request.user until process_response. This makes it available in process_view.
-            # For more info, see https://github.com/jpadilla/django-rest-framework-jwt/issues/45#issuecomment-74996698
-            request.user = SimpleLazyObject(lambda: _get_user_from_jwt(request, view_func))
+            # Since this call to the user is not made lazily, and has the potential to cause issues, we
+            # ensure it is only used in the case of ENABLE_SET_REQUEST_USER_FOR_JWT_COOKIE.
+            if not get_user(request).is_authenticated:
+                # Similar to django/contrib/auth/middleware.py AuthenticationMiddleware.
+                set_flag_is_request_user_set_for_jwt_auth()
+                request.user = SimpleLazyObject(lambda: _get_cached_user_from_jwt(request, view_func))
+
+
+def _get_module_request_cache():
+    return RequestCache(__name__).data
+
+
+def _get_cached_user_from_jwt(request, view_func):
+    """
+    Returns cached user from JWT authentication.
+
+    Performs JWT authentication if not already cached.
+    """
+    # Similar to django/contrib/auth/middleware.py get_user.
```

Comment:
**Note to reviewer:** I know this layer of caching may not be necessary, since we don't yet have it. But it seemed like the implementation should be more parallel to what exists in the Django middleware.

@feanil: Feel free to comment or mark resolved if you have no comment.

Reply:
This makes sense, I'll resolve after the review.